org.dbdoclet.herold.package/doc/manpage/herold.1
'\" t
.\"     Title: herold
.\"    Author: Michael Fuchs
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 11/21/2015
.\"    Manual: User Commands
.\"    Source: herold
.\"  Language: English
.\"
.TH "HEROLD" "1" "11/21/2015" "herold" "User Commands"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
herold \- HTML to DocBook converter
.SH "SYNOPSIS"
.HP \w'\fBherold\fR\ 'u
\fBherold\fR [OPTIONS]
.SH "DESCRIPTION"
.PP
The reuse of HTML content in presentation\-neutral form is a frequent problem\&. One possible solution is to convert HTML to DocBook XML, because DocBook is a semantic markup language for documentation, which enables its users to create document content that captures the logical structure of the content\&.
.PP
The command line tool
herold
can be used to convert HTML to DocBook\&. Because HTML elements are often used not as intended, the possibilities for such a transformation are somewhat limited\&. herold is part of the dbdoclet suite of tools\&. For more information visit
\m[blue]\fBhttp://www\&.dbdoclet\&.org\fR\m[]\&.
.SH "OPTIONS"
.PP
\-\-docbook\-add\-index, \-x
.RS 4
Automatically add an index element at the end of the document\&.
.RE
.PP
\-\-docbook\-decompose\-tables, \-T
.RS 4
Decomposes the tables from the HTML code into single paragraphs\&. This can be useful, if a document contains a lot of tables for formatting reasons\&.
.RE
.PP
\-\-docbook\-encoding, \-d
.RS 4
Specifies the encoding of the generated DocBook XML files\&.
.RE
.PP
\-\-docbook\-root\-element, \-r
.RS 4
The root element of the document\&. Possible values are: book, article, reference, part, chapter or section\&. The default value for this option is \*(Aqarticle\*(Aq
.RE
.PP
\-\-docbook\-title, \-t
.RS 4
The title for the resulting document\&.
.RE
.PP
\-\-in, \-i
.RS 4
Specifies the HTML input file\&.
.RE
.PP
\-\-help, \-h
.RS 4
Prints a help page on the console\&.
.RE
.PP
\-\-html\-encoding, \-s
.RS 4
Specifies the encoding of the HTML source files, such as ISO\-8859\-1\&.
.RE
.PP
\-\-out, \-o
.RS 4
Specifies the DocBook XML destination file\&.
.RE
.PP
\-\-profile, \-p
.RS 4
A profile file with predefined settings\&.
.RE
.PP
\-\-verbose, v
.RS 4
Enables the verbosity for the console output\&.
.RE
.PP
\-\-version, \-V
.RS 4
Displays the version of herold\&.
.RE
.SH "CONFIGURATION"
.PP
The details of a transformation are controlled by a profile file\&. A profile file offers more possibilities to influence the transformation than the command line arguments\&. The following example shows a typical profile file\&.
.sp
.if n \{\
.RS 4
.\}
.nf
transformation html2docbook;

section section\-detection  {
    attribute\-class = ["^MsoHeading(\ed+)$"];
    section\-numbering\-pattern = "((\ed+\e\&.)+)?\ed*\e\&.?\ep{Z}*";
}

section list\-detection {
    itemized\-attribute\-class = ["^MsoListBullet(\ew*)$", "Aufzhlung(\ew+)$];
    itemized\-strip\-prefix = [ "\-", "o", "\eu00b7" ];
    ordered\-attribute\-class = ["^MsoListNumbered(\ew*)$"];
    ordered\-strip\-prefix = [ "\ed+\e\&.\es+" ];
}

section HTML {
    encoding = "windows\-1252";
    exclude = [ "//p[starts\-with(@class, \*(AqMsoToc\*(Aq)]", "" ];
}

section DocBook {
    abstract = """<title>Lorem ipsum</title>
<para>Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed 
do eiusmod tempor incididunt ut labore et dolore magna aliqua\&. Ut 
enim ad minim veniam, quis nostrud exercitation ullamco laboris 
nisi ut aliquip ex ea commodo consequat\&. Duis aute irure dolor in 
reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
pariatur\&. Excepteur sint occaecat cupidatat non proident, sunt in
culpa qui officia deserunt mollit anim id est laborum\&.sed, dolor 
amet\&.</para>""";
    add\-index = true;
    author\-email = "me@somewhere\&.de";
    author\-firstname = "Michael";
    author\-surname = "Fuchs";
    chunk\-elements = [ "chapter", "section", "appendix" ];
// Syntax: chunk\-<CHUNK\-ELEMENT>\-depth = <INT>;
    chunk\-section\-depth = 3;  
    collapse\-protected\-space = "true";
    copyright\-holder = "Ingenieurbüro Michael Fuchs";
    copyright\-year = "2015";
    corporation = "";
    create\-condition\-attribute = false;
    create\-prolog = true;
    create\-remap\-attribute = false;
    create\-xref\-label = false;
    decompose\-tables = false;
    detect\-trapped\-br = true;
    documentation\-id = "doc01";
    document\-element = "book";
    encoding = "UTF\-8";
    hyphenation\-char = "soft\-hyphen";
    image\-data\-formats = [ "gif", "base64" ];
    image\-path = "\&./figures";
    language = "de";
    release\-info = "Version 3\&.1";
    table\-style = "all";
    title = "Tutorial";
    title\-normalize\-space = true;
    use\-absolute\-image\-path = false;
}
.fi
.if n \{\
.RE
.\}
.SS "Syntax"
.PP
A profile file consists mainly of sections\&. Sections are used to group parameters which share the same context\&. Every section must start with the keyword
\fIsection\fR
followed by the name of the section\&. After the name comes the block of parameters, which is surrounded by curly braces\&. Parameters can be of type String, Number, Boolean or Array\&. Strings must be framed with double quotes\&. If the String contains newlines, use three double quotes instead of one\&. Arrays are framed with square brackets\&. Inside an array, the elements must be comma separated\&. Every assignment must be finished by a semicolon\&. Multi line comments have the form
\fI/* my comment */\fR
, single line comments look like
\fI// my comment\en\fR\&.
.SS "Mandatory Elements"
.PP
A profile for herold must start with the line
transformation html2docbook;\&.
.SS "Section HTML"
.PP
The section HTML defines parameters, which control the loading and parsing of the HTML input data\&.
.PP
.PP
encoding
.RS 4
The character set used to read the input stream\&.
.RE
.PP
exclude
.RS 4
Defines an array of xpath expressions\&. All matches are removed from the HTML DOM tree before transformation\&.
.RE
.SS "Section DocBook"
.PP
.PP
abstract
.RS 4
The text for the abstract element of the info section\&. If the text is structured with newlines, use three double quotes as delimiters\&. If the text starts with a "<" character, it is embedded into an abstract element, otherwise the text is embedded into an para element inside of an abstract element\&. The text will parsed and can contain DocBook elements\&.
.RE
.PP
add\-index
.RS 4
If set to true, an index element is inserted at the end of the DocBook XML\&.
.RE
.PP
author\-email
.RS 4
The email address of the author\&. If this parameter is set, it is used to create an info section at the beginning of the document\&.
.RE
.PP
author\-firstname
.RS 4
The firstname of the author\&. If this parameter is set, it is used to create an info section at the beginning of the document\&.
.RE
.PP
author\-surname
.RS 4
The surname of the author\&. If this parameter is set, it is used to create an info section at the beginning of the document\&.
.RE
.PP
chunk\-elements
.RS 4
Defines an array of element names\&. If an element of this list is detected while writing the output, the element and all child nodes will be written to a separate file\&. This new file will be included into the parent file with an
\fBxi:include\fR
tag\&. Recursive structures result in recursive includes\&. You might want to use this, if you are transforming big HTML files and the resulting DocBook XML file becomes uncomfortable large\&.
.RE
.PP
chunk\-<CHUNK\-ELEMENT>\-depth
.RS 4
Defines the depth for a chunk element, until the chunking should be executed, eg
chunk\-section\-depth = 3\&. If an element defined for chunking is nested recursivley, you might want to control the depth to which the chunking should be done\&. The default depth is 1, which means only the topmost element is separated\&.
.RE
.PP
create\-xref\-label
.RS 4
if set to false, anchor elements doesn\*(Aqt get a xreflabel attribute\&.
.RE
.PP
decompose\-tables
.RS 4
If set to true, tables structures will be ignored\&. The content of the table cells will be inserted into the DocBook XML as a sequence of paragraphs\&. This parameter can be useful if your HTML contains tables for formatting purposes\&. Normally you want to get rid of them, because they tamper the logical structure\&.
.RE
.PP
document\-element
.RS 4
The document element you want to use\&. Must be one of article, book, part or reference\&.
.RE
.PP
encoding
.RS 4
The character set which will be used for writing the output file\&.
.RE
.PP
image\-data\-formats
.RS 4
An array of image formats\&. These formats will be inserted as imageobject elements, additionally to the format found in the src attribute of the corresponding img element\&. The original format is inserted twice with the roles "html" and "fo"\&. The other formats are inserted as "html\-<FORMAT>" and "fo\-<FORMAT>"\&.
.RE
.PP
title
.RS 4
The title of the resulting document\&. If this parameter is undefined, herold tries to dected the title from the head section of the HTML data\&.
.RE
.PP
use\-absolute\-image\-path
.RS 4
If you want absolute image paths in the fileref attribute of the imagedata element, set this parameter to true\&.
.RE
.SS "Section node"
.PP
The mapping of HTML elements to DocBook element can be fine tuned by using node sections\&. If you have HTML code which looks like the following fragment:
.sp
.if n \{\
.RS 4
.\}
.nf
<ol class="procedure">
  <li>Step 1</li>
  <li>Step 2</li>
  <li>Step 3</li>
</ol>
.fi
.if n \{\
.RE
.\}
.sp
The resulting DocBook XML after the transformation would normally look like:
.sp
.if n \{\
.RS 4
.\}
.nf
<orderedlist>
  <listitem>Step 1</listitem>
  <listitem>Step 2</listitem>
  <listitem>Step 3</listitem>
</orderlist>
.fi
.if n \{\
.RE
.\}
.sp
But what you would like to have is something like:
.sp
.if n \{\
.RS 4
.\}
.nf
<procedure>
  <step>Step 1</step>
  <step>Step 2</step>
  <step>Step 3</step>
</procedure>
.fi
.if n \{\
.RE
.\}
.sp
To achieve this, you can use the following rules in our profile:
.sp
.if n \{\
.RS 4
.\}
.nf
node "//ol[@class=\*(Aqprocedure\*(Aq]" {
  map\-to = "procedure";
}

node "//ol[@class=\*(Aqprocedure\*(Aq]/li" {
  map\-to = "step";
}
.fi
.if n \{\
.RE
.\}
.sp
After the keyword
node
follows a xpath expression which is matched against the document element of the HTML file (typically <html>)\&. The parameter
map\-to
defines the DocBook element, which is used instead of the default mapping element\&.
.SS "Section attribute"
.PP
An attribute section is more or less the same as a node section\&. Instead of redefining the mapping of a HTML element to a DocBook element, the mapping for an attribute is changed\&. The following section maps an attribute
class=\*(Aqprocedure\*(Aq
to
role=\*(Aqprocedure\*(Aq\&.
.sp
.if n \{\
.RS 4
.\}
.nf
attribute "//@class[contains(\&., \*(Aqprocedure\*(Aq)]" {
	map\-to = "role";
}
      
.fi
.if n \{\
.RE
.\}
.sp
.SS "Section section\-detection"
.PP
The section
\fIsection\-detection\fR
is used to detect section elements in HTML code and to strip off any numbering prefix from the titles\&.
.PP
Many authoring tools allow deeply nested sections\&. While exporting HTML, it happens, that the nesting becomes deeper than six levels\&. HTML provides header elements for up to six levels, h1\-h6, but no h7 or even more\&. At this point, the formatting is normally done with the help of CSS and div or p elements\&. herold is able to detect the header element of HTML, but it can not know about the export format of a specific tool\&. To solve this problem even for some cases, you can specify the parameter
\fIattribute\-class\fR\&. It consists of a list of regular expressions, which are matched against the class attribute of each HTML element\&. If a match is found, the element is considered as a section element\&. The regular expression can have group, which is interpreted as level indicator\&. The group must be the first group and it must match against a number, e\&.g\&.
^heading(\ed+)$\&. If the level can not be detected, a level of seven is assumed\&.
.PP
Because DocBook XSL stylesheets take care of the section numbering while transforming the DocBook XML to a specific output, it is often necessary to strip the numbering already defined in the HTML page\&. Otherwise you end up with two numbering texts in front of your titles\&. To help herold with the detection of numbering patterns, use the parameter
\fIsection\-numbering\-pattern\fR\&.
.PP
.PP
attribute\-class
.RS 4
A regular expression, which is applied to every p and div element\&. If the expression matches, the current element is handled as a section element\&. If the regular expression has groups, the first group will be used as nesting level, otherwise level seven is assumed\&.
.RE
.PP
section\-numbering\-pattern
.RS 4
Normally you want to get rid of the section numbering that comes with the HTML data, because it becomes part of the title text in DocBook\&. The section numbers will the appear twice in your target media\&. One from HTML and one from the DocBook XSL processing\&. The parameter section\-numbering\-pattern defines a regular expression, which is matched against the beginning of every section title\&. If it matches, the matching part is removed\&.
.RE
.SS "Section list\-detection"
.PP
Sometimes lists are not represented with ul, ol or dl tags, but they are represented as p tags with additional css formatting\&. If you use a tool, which creates or exports HTML with such a construct, the conversion will end up with para elements, instead of the corresponding list elements in DocBook\&. To recreate the lists in some cases, you can use the section
\fIlist\-detection\fR\&. The parameters
\fIitemized\-attribute\-class\fR
and
\fIordered\-attribute\-class\fR
let you define lists of regular expression, which should match against the class attribute of listitem elements in the HTML\&. herold tries to rebuild the proper list structure from this information, even for nested lists\&.
.SH "COPYRIGHT"
.PP
Copyright 2001\-2015 Michael Fuchs\&. License GPLv3+: GNU GPL version 3 or later
\m[blue]\fBhttp://gnu\&.org/licenses/gpl\&.html\fR\m[]\&. This is free software: you are free to change and redistribute it\&. There is NO WARRANTY, to the extent permitted by law\&.
.SH "AUTHOR"
.PP
\fBMichael Fuchs\fR
.RS 4
Software Engineer
.RE


org.dbdoclet.herold.package/src/main/java/org/dbdoclet/herold/Herold.java
/* 
 * ### Copyright (C) 2008 Michael Fuchs ###
 * ### All Rights Reserved.             ###
 *
 * Author: Michael Fuchs
 * E-Mail: michael.fuchs@unico-group.com
 * URL:    http://www.michael-a-fuchs.de
 */
package org.dbdoclet.herold;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.StringWriter;
import java.net.URL;
import java.text.MessageFormat;
import java.util.ResourceBundle;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.dbdoclet.option.BooleanOption;
import org.dbdoclet.option.FileOption;
import org.dbdoclet.option.Option;
import org.dbdoclet.option.OptionException;
import org.dbdoclet.option.OptionList;
import org.dbdoclet.option.SelectOption;
import org.dbdoclet.option.StringOption;
import org.dbdoclet.service.FileServices;
import org.dbdoclet.service.ResourceServices;
import org.dbdoclet.service.StringServices;
import org.dbdoclet.trafo.AbstractTrafoService;
import org.dbdoclet.trafo.TrafoConstants;
import org.dbdoclet.trafo.TrafoException;
import org.dbdoclet.trafo.TrafoExceptionHandler;
import org.dbdoclet.trafo.TrafoResult;
import org.dbdoclet.trafo.TrafoScriptManager;
import org.dbdoclet.trafo.html.dita.HtmlDitaTrafo;
import org.dbdoclet.trafo.html.docbook.HtmlDocBookTrafo;
import org.dbdoclet.trafo.param.BooleanParam;
import org.dbdoclet.trafo.param.TextParam;
import org.dbdoclet.trafo.script.Namespace;
import org.dbdoclet.trafo.script.Script;
import org.dbdoclet.trafo.script.Section;

/**
 * The class <code>Herold</code> implements a console converter to transform
 * HTML code to DocBook XML.
 * 
 * @author <a href ="mailto:mfuchs@unico-consulting.com">Michael Fuchs</a>
 * @version 1.0
 */
public class Herold {

	public enum OutputFormat {
		DITA, DocBook
	}

	private static Log logger = LogFactory.getLog(Herold.class);;

	private int exitCode;
	private FileOption optIn;
	private FileOption optOut;
	private OutputFormat outputFormat;
	private ResourceBundle res = ResourceBundle
			.getBundle("org/dbdoclet/herold/Resources");
	private boolean systemExitEnabled = true;
	private boolean verbose = false;

	private static String getVersion() {

		Package p = Herold.class.getPackage();
		return p.getImplementationVersion();
	}

	public static void main(String[] args) {

		Herold herold = new Herold();
		herold.execute(args);
	}

	private static void printLicense() throws IOException {

		URL url = ResourceServices
				.getResourceAsUrl("/org/dbdoclet/herold/COPYING");

		if (url != null) {
			BufferedReader reader = new BufferedReader(new InputStreamReader(
					url.openStream()));
			String line = reader.readLine();

			while (line != null) {
				println(line);
				line = reader.readLine();
			}

			reader.close();
		} else {
			logger.fatal("Can't find resource for license!");
		}
	}

	private static void println(String str) {
		System.out.println(str);
	}

	private static void printVersion() {

		println("herold version \"" + getVersion() + "\"");
	}

	public void convert(File htmlFile, File xmlFile) throws TrafoException,
			FileNotFoundException {

		Script script = new Script();
		convert(new FileInputStream(htmlFile), new FileOutputStream(xmlFile), xmlFile,
				script);
	}

	public void convert(InputStream in, OutputStream out, File systemId, Script script)
			throws TrafoException {

		AbstractTrafoService trafo = new HtmlDocBookTrafo();

		if (outputFormat == OutputFormat.DITA) {
			trafo = new HtmlDitaTrafo();
		}

		trafo.setSystemId(systemId);
		trafo.setInputStream(in);
		trafo.setOutputStream(out);

		TrafoResult result = null;
		if (verbose == true) {
			trafo.addProgressListener(new ConsoleProgressListener(false));
			result = trafo.transform(script);
		} else {
			result = trafo.transform(script);
		}

		if (result.isFailed()) {
			System.err.print(result.toString());
		}

		if (verbose == true) {
			System.out.println();
		}
	}

	private OptionList createOptionList(String[] args) {

		Option<?> option;
		SelectOption selopt;
		BooleanOption bopt;
		StringOption sopt;

		OptionList options = new OptionList(args);

		// docbook-add-index
		bopt = new BooleanOption(TrafoConstants.SECTION_DOCBOOK.toLowerCase()
				+ "-" + TrafoConstants.PARAM_ADD_INDEX, "x");
		bopt.setValue(false);
		options.add(bopt);

		// docbook-encoding
		sopt = new StringOption(TrafoConstants.SECTION_DOCBOOK.toLowerCase()
				+ "-" + TrafoConstants.PARAM_ENCODING, "d");
		sopt.setDefault("UTF-8");
		options.add(sopt);

		// decompose-tables
		bopt = new BooleanOption(TrafoConstants.SECTION_DOCBOOK.toLowerCase()
				+ "-" + TrafoConstants.PARAM_DECOMPOSE_TABLES, "T");
		bopt.setDefault(false);
		options.add(bopt);

		// document-element
		selopt = new SelectOption(TrafoConstants.SECTION_DOCBOOK.toLowerCase()
				+ "-" + TrafoConstants.PARAM_DOCUMENT_ELEMENT, "r");

		String[] optv2 = { "article", "book", "reference", "part", "chapter",
				"section" };
		selopt.setList(optv2);
		selopt.setDefault("article");
		options.add(selopt);

		// title
		sopt = new StringOption(TrafoConstants.SECTION_DOCBOOK.toLowerCase()
				+ "-" + TrafoConstants.PARAM_TITLE, "t");
		sopt.setDefault("http://www.dbdoclet.org/herold");
		options.add(sopt);

		// source-encoding
		sopt = new StringOption(TrafoConstants.SECTION_HTML.toLowerCase() + "-"
				+ TrafoConstants.PARAM_ENCODING, "s");
		sopt.setDefault("UTF-8");
		options.add(sopt);

		// in
		optIn = new FileOption("in", "i");
		optIn.isExisting(true);
		options.add(optIn);

		// loglevel
		SelectOption optLogLevel = new SelectOption("logging-level", "l");
		optLogLevel.setList(new String[] { "debug5", "debug4", "debug3",
				"debug2", "debug", "info", "warn", "error", "fatal" });
		optLogLevel.setDefault("error");
		options.add(optLogLevel);

		// out
		optOut = new FileOption("out", "o");
		options.add(optOut);

		StringOption optProfile = new StringOption("profile", "p");
		options.add(optProfile);

		// verbose
		option = new BooleanOption("verbose", "v");
		options.add(option);

		return options;
	}

	public OutputFormat getOutputFormat() {
		return outputFormat;
	}

	public void setOutputFormat(OutputFormat outputFormat) {
		this.outputFormat = outputFormat;
	}

	private OutputStream createOutputStream() throws FileNotFoundException {

		OutputStream out;

		if (optOut.isUnset()) {

			out = System.out;

		} else {

			File outDir = optOut.getValue().getParentFile();

			if (outDir != null && outDir.exists() == false) {
				outDir.mkdir();
			}

			out = new FileOutputStream(optOut.getValue());
		}

		return out;
	}

	public void execute(String[] args) {

		OptionList options = null;
		exitCode = 0;

		try {

			Option<?> option;
			options = new OptionList(args);

			// help
			option = new BooleanOption("help", "h");
			option.setMediumName("?");
			options.add(option);

			// license
			option = new BooleanOption("license", "L");
			options.add(option);

			// version
			option = new BooleanOption("version", "V");
			options.add(option);

			options.validate(true);

			if (options.getFlag("help", false)) {
				printUsage();
				return;
			}

			if (options.getFlag("version", false)) {
				printVersion();
				return;
			}

			if (options.getFlag("license", false)) {
				printLicense();
				exit(0);
			}

			options = createOptionList(args);

			if (options.validate() == false) {
				throw new OptionException(options.getError());
			}

			StringOption optProfile = (StringOption) options
					.getOption("profile");
			File profileFile = findProfile(optProfile);
			Script script;

			TrafoScriptManager mgr = new TrafoScriptManager();
			script = mgr.parseScript(profileFile);
			logger.info(String.format("Using profile file %s.",
					profileFile.getAbsolutePath()));
			
			if (logger.isTraceEnabled()) {
				StringWriter buffer = new StringWriter();
				mgr.writeScript(script, buffer);
				logger.trace("Script: " + buffer.toString());
			}

			File outFile = processCommandLineOptions(options, script);

			if (outFile != null) {
				logger.info(String.format("Output file is %s.",
						outFile.getPath()));
			}

			InputStream in = null;
			OutputStream out = null;

			if (optIn.isUnset()) {
				in = System.in;
			} else {
				in = new FileInputStream(optIn.getValue());
			}

			out = createOutputStream();

			convert(in, out, optOut.getValue(), script);

		} catch (OptionException oops) {

			if ((options != null) && (options.getFlag("help", false) == false)) {

				printUsage();

				String msg = oops.getMessage();

				if (msg != null) {
					System.err.println(msg);
				}

			} else {
				oops.printStackTrace();
				printUsage();
			}

			exitCode = 1;

		} catch (TrafoException oops) {

			logger.fatal("TrafoException", TrafoExceptionHandler.getCause(oops));
			exitCode = 2;

		} catch (FileNotFoundException oops) {

			logger.fatal(oops.getMessage());
			exitCode = 3;

		} catch (IOException oops) {

			oops.printStackTrace();
			exitCode = 4;

		} catch (Exception oops) {

			oops.printStackTrace();
			exitCode = 5;
		}

		exit(exitCode);
	}

	private void exit(int exitCode) {

		this.exitCode = exitCode;

		if (systemExitEnabled == true) {
			System.exit(exitCode);
		}
	}

	private File findProfile(StringOption optProfile)
			throws FileNotFoundException {

		String profileName = "default.her";

		if (optProfile.isUnset() == false) {
			profileName = optProfile.getValue();
		}

		if (profileName == null || profileName.trim().length() == 0) {
			profileName = "default.her";
		}

		File profile = new File(profileName);

		if (profile.exists() == true) {
			return profile;
		}

		String homePath = System.getProperty("herold.home");

		if (homePath != null) {

			File homeDir = new File(homePath);

			if (homeDir.exists() == false) {
				throw new FileNotFoundException(String.format(
						"Home directory %s does not exist!", homePath));
			}

			File profilesDir = new File(homeDir, "profiles");

			if (profilesDir.exists() == false) {
				throw new FileNotFoundException(String.format(
						"Profiles directory %s does not exist!",
						profilesDir.getAbsolutePath()));
			}

			File file = new File(profilesDir, profileName);

			if (file.exists() == true) {
				return file;
			}

			file = new File(profilesDir, profileName + ".her");

			if (file.exists() == true) {
				return file;
			}
		}

		throw new FileNotFoundException(String.format(
				"Profile %s does not exist!", profileName));
	}

	public int getExitCode() {
		return exitCode;
	}

	public boolean isCanceled() {
		return false;
	}

	private void printUsage() {

		try {

			String resname = ResourceServices.getString(res, "C_USAGE");
			String buffer = ResourceServices.getResourceAsString(resname);

			println(MessageFormat.format(buffer, getVersion()));

		} catch (IOException oops) {

			logger.fatal("Printing usage message failed!", oops);
		}
	}

	public File processCommandLineOptions(OptionList options, Script script) {

		Namespace namespace = script.getNamespace();
		Section section = null;
		
		for (Option<?> option : options) {

			String name = option.getLongName();

			if (name.startsWith(TrafoConstants.SECTION_DOCBOOK.toLowerCase())) {
				
				section = namespace.findOrCreateSection(TrafoConstants.SECTION_DOCBOOK);
				name = StringServices.cutPrefix(name,
						TrafoConstants.SECTION_DOCBOOK.toLowerCase());

			} else if (name.startsWith(TrafoConstants.SECTION_HTML
					.toLowerCase())) {

				section = namespace.findOrCreateSection(TrafoConstants.SECTION_HTML);
				name = StringServices.cutPrefix(name,
						TrafoConstants.SECTION_HTML.toLowerCase());
			}

			name = StringServices.cutPrefix(name, "-");

			if (section != null && option.isUnset() == false) {
				switch (option.getType()) {

				case BOOLEAN:
					section.setParam(new BooleanParam(name, (Boolean) option.getValue()));
					break;

				default:
					section.setParam(new TextParam(name, option.getValue().toString()));
					break;
				}
			}
		}

		setVerbose(options.getFlag("verbose", false));

		File outFile;

		if (optOut == null) {
			throw new IllegalStateException(
					"The field optOut must not be null!");
		}

		if (optIn == null) {
			throw new IllegalStateException(
					"The field optIn must not be null!");
		}

		if (optOut.isUnset() && optIn.isUnset() == false) {

			String outFileName = FileServices.getFileBase(optIn.getValue())
					+ ".xml";
			// outFile = FileServices.createUniqueFile(new File(outFileName));
			outFile = new File(outFileName);

		} else {
			outFile = optOut.getValue();
		}

		return outFile;
	}

	public void setSystemExitEnabled(boolean systemExitEnabled) {
		this.systemExitEnabled = systemExitEnabled;
	}

	public void setVerbose(boolean verbose) {
		this.verbose = verbose;
	}
}
